src/node/redeem.rs

```diff
@@ -0,0 +1,191 @@
+// Rust Simplicity Library
+// Written in 2023 by
+//   Andrew Poelstra <apoelstra@blockstream.com>
+//
+// To the extent possible under law, the author(s) have dedicated all
+// copyright and related and neighboring rights to this software to
+// the public domain worldwide. This software is distributed without
+// any warranty.
+//
+// You should have received a copy of the CC0 Public Domain Dedication
+// along with this software.
+// If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+//
+
+use crate::analysis::NodeBounds;
+use crate::dag::MaxSharing;
+use crate::jet::Jet;
+use crate::types::{self, arrow::FinalArrow};
+use crate::{Amr, Cmr, FirstPassImr, Imr, Value};
+
+use super::{CommitData, CommitNode, Inner, NoWitness, Node, NodeData};
+
+use std::marker::PhantomData;
+use std::sync::Arc;
+
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+pub enum Redeem {}
+
+impl<J: Jet> NodeData<J> for Redeem {
+    type CachedData = Arc<RedeemData<J>>;
+    type Witness = Arc<Value>;
+    type SharingId = (Cmr, Imr);
```

Comment:
1778550599030115c2d33b7417d46aa9e7e734f2: Let's make this `type SharingId = Imr`